ValveResourceFormat/Serialization/VfxEval.cs

```diff
@@ -0,0 +1,503 @@
+            }
+
+            if (op == OPCODE.NEGATE)
+            {
+                var exp = Expressions.Pop();
+                Expressions.Push($"-{exp}");
+                return;
+            }
+
+            if (op == OPCODE.EXTVAR)
+            {
+                var varId = dataReader.ReadUInt32();
+                var extVarname = GetExternalVarName(varId);
+                Expressions.Push(extVarname);
+                return;
+            }
+
+            if (op == OPCODE.SWIZZLE)
+            {
+                var exp = Expressions.Pop();
+                exp += $".{GetSwizzle(dataReader.ReadByte())}";
+                Expressions.Push($"{exp}");
+                return;
+            }
+
+            if (op == OPCODE.EXISTS)
+            {
+                var varId = dataReader.ReadUInt32();
+                var extVarname = GetExternalVarName(varId);
+                Expressions.Push($"exists({extVarname})");
+                return;
+            }
+
+            // parser terminates here
+            if (op == OPCODE.ENDOFDATA)
+            {
+                if (dataReader.PeekChar() != -1)
+                {
+                    ErrorMessage = "malformed data!";
+                    ErrorWhileParsing = true;
+                    return;
+                }
+                var finalExp = Expressions.Pop();
+                finalExp = Trimb(finalExp);
+                DynamicExpressionList.Add($"return {finalExp};");
+                return;
+            }
+
+            // this point should never be reached
+            ErrorWhileParsing = true;
+            ErrorMessage = $"UNKNOWN OPCODE = 0x{(int)op:x2}, offset = {dataReader.BaseStream.Position}";
+        }
+
+        private void ApplyFunction(string funcName, int nrArguments)
+        {
+            if (nrArguments == 0)
+            {
+                Expressions.Push($"{funcName}()");
+                return;
+            }
+            var exp1 = Expressions.Pop();
+            if (nrArguments == 1)
+            {
+                Expressions.Push($"{funcName}({Trimb(exp1)})");
+                return;
+            }
+            var exp2 = Expressions.Pop();
+            if (nrArguments == 2)
+            {
+                Expressions.Push($"{funcName}({Trimb(exp2)},{Trimb(exp1)})");
+                return;
+            }
+            var exp3 = Expressions.Pop();
+            if (nrArguments == 3)
+            {
+                // trim or not to trim ...
+                Expressions.Push($"{funcName}({Trimb(exp3)},{Trimb(exp2)},{Trimb(exp1)})");
+                // expressions.Push($"{funcName}({exp3},{exp2},{exp1})");
+                return;
+            }
+            var exp4 = Expressions.Pop();
+            if (nrArguments == 4)
+            {
+                Expressions.Push($"{funcName}({Trimb(exp4)},{Trimb(exp3)},{Trimb(exp2)},{Trimb(exp1)})");
+                return;
+            }
+
+            throw new Exception("this cannot happen!");
```

Comment:
```suggestion
            throw new Exception($"Dynamic expression: Unexpected number of arguments ({nrArguments}) for function ${funcName}");
```

Reply:
I made this change, this is primarily a developer note because it really cannot happen, the value for nrArguments passed to ApplyFunction(..) is taken from the FUNCTION_REF lookup which only has values in the range [0,4]

Reply:
It cannot happen until it does. And when it does, the message should be reasonable :)